weave/flow/eval_imperative.py

```diff
@@ -0,0 +1,358 @@
+def _set_current_score(score: ScoreType) -> Iterator[None]:
+    token = current_score.set(score)
+    try:
+        yield
+    finally:
+        current_score.reset(token)
+
+
+@contextmanager
+def _set_current_summary(summary: dict) -> Iterator[None]:
+    token = current_summary.set(summary)
+    try:
+        yield
+    finally:
+        current_summary.reset(token)
+
+
+def _cast_to_cls(type_: type[T]) -> Callable[[str | T], T]:
+    def _convert_to_cls_inner(value: str | T) -> T:
+        if isinstance(value, str):
+            cls_name = value
+
+            # Dynamically create the class if the user only provides a name
+            cls_name = _validate_class_name(cls_name)
+
+            pydantic_config_dict = {
+                "__annotations__": {"name": str},
+                "name": cls_name,
+            }
+
+            cls = type(cls_name, (type_,), pydantic_config_dict)
+            return cast(T, cls())
+
+        elif isinstance(value, dict):
+            attributes = value
+
+            # Dynamically create the class with attributes from the dict)
+            cls_name = f"Dynamic{type_.__name__}"
+            if "name" not in attributes:
+                attributes["name"] = cls_name
+
+            pydantic_config_dict = {
+                "__annotations__": dict.fromkeys(attributes, Any),
+                **attributes,
+            }
+            cls = type(cls_name, (type_,), pydantic_config_dict)
+            return cast(T, cls())
+
+        elif isinstance(value, type_):
+            return value
+
+        raise TypeError("Unsupported type for casting")
+
+    return _convert_to_cls_inner
+
+
+def _validate_class_name(name: str) -> str:
+    """Validate the scorer name to be a valid class name."""
+    # Check if name is not empty
+    if not name:
+        raise ValueError("Scorer name cannot be empty")
+
+    # Check if name follows Python class naming conventions
+    # Class names should start with a letter or underscore and contain only
+    # alphanumeric characters and underscores
+    import re
+
+    if not re.match(r"^[a-zA-Z_][a-zA-Z0-9_]*$", name):
+        raise ValueError(
+            f"Invalid scorer name: '{name}'. Scorer names must start with a letter or underscore "
+            "and contain only alphanumeric characters and underscores."
+        )
+
+    # Check if name is not a Python keyword
+    import keyword
+
+    if keyword.iskeyword(name):
+        raise ValueError(f"Scorer name '{name}' cannot be a Python keyword")
+
+    return name
+
+
+class ImperativeScoreLogger(BaseModel):
+    """This class provides an imperative interface for logging scores."""
+
+    # model_id: ID
+    predict_and_score_call: Call
+    evaluate_call: Call
+    predict_call: Call
+
+    _captured_scores: dict[str, ScoreType] = PrivateAttr(default_factory=dict)
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def finish(self, *, scores: dict[str, ScoreType] | None = None) -> None:
```

Comment:
Similar to "summarize" should we disallow adding scores after finish?

Reply:
yes